tests/integration/staking/keeper/msg_server_test.go

```diff
@@ -172,3 +177,241 @@ func TestCancelUnbondingDelegation(t *testing.T) {
+			expHistoryObjs: 1,
+			fees:           keyRotationFee,
+		},
+		{
+			name: "non existing validator check",
+			malleate: func() sdk.Context {
+				return ctx
+			},
+			validator: sdk.ValAddress("non_existing_val"),
+			newPubKey: PKs[498],
+			pass:      false,
+			expErrMsg: "validator does not exist",
+		},
+		{
+			name: "pubkey already associated with another validator",
+			malleate: func() sdk.Context {
+				return ctx
+			},
+			validator: validators[0].GetOperator(),
+			newPubKey: validators[1].ConsensusPubkey.GetCachedValue().(cryptotypes.PubKey),
+			pass:      false,
+			expErrMsg: "consensus pubkey is already used for a validator",
+		},
+		{
+			name: "consensus pubkey rotation limit check",
+			malleate: func() sdk.Context {
+				params := stakingKeeper.GetParams(ctx)
+				params.KeyRotationFee = sdk.NewInt64Coin(bondDenom, 10)
+				params.MaxConsPubkeyRotations = 1
+				err := stakingKeeper.SetParams(ctx, params)
+				require.NoError(t, err)
+
+				msg, err := types.NewMsgRotateConsPubKey(
+					validators[1].GetOperator(),
+					PKs[498],
+				)
+				require.NoError(t, err)
+				_, err = msgServer.RotateConsPubKey(ctx, msg)
+				require.NoError(t, err)
+
+				return ctx
+			},
+			validator: validators[1].GetOperator(),
+			newPubKey: PKs[497],
+			pass:      false,
+			expErrMsg: "exceeding maximum consensus pubkey rotations within unbonding period",
+		},
+		{
+			name: "two rotations within unbonding period",
+			malleate: func() sdk.Context {
+				params := stakingKeeper.GetParams(ctx)
+				params.KeyRotationFee = sdk.NewInt64Coin(bondDenom, 10)
+				params.MaxConsPubkeyRotations = 2
+				err := stakingKeeper.SetParams(ctx, params)
+				require.NoError(t, err)
+
+				msg, err := types.NewMsgRotateConsPubKey(
+					validators[2].GetOperator(),
+					PKs[497],
+				)
+				require.NoError(t, err)
+				_, err = msgServer.RotateConsPubKey(ctx, msg)
+				require.NoError(t, err)
+				ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+
+				return ctx
+			},
+			validator:      validators[2].GetOperator(),
+			newPubKey:      PKs[496],
+			pass:           true,
+			fees:           calculateFee(keyRotationFee, 1),
+			expHistoryObjs: 2,
+		},
+		{
+			name: "limit reached, but should rotate after the unbonding period",
+			malleate: func() sdk.Context {
+				params := stakingKeeper.GetParams(ctx)
+				params.KeyRotationFee = sdk.NewInt64Coin(bondDenom, 10)
+				params.MaxConsPubkeyRotations = 1
+				err := stakingKeeper.SetParams(ctx, params)
+				require.NoError(t, err)
+
+				msg, err := types.NewMsgRotateConsPubKey(
+					validators[3].GetOperator(),
+					PKs[495],
+				)
+
+				require.NoError(t, err)
+				_, err = msgServer.RotateConsPubKey(ctx, msg)
+				require.NoError(t, err)
+				ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+
+				// this shouldn't remove the existing keys from waiting queue since unbonding time isn't reached
+				stakingKeeper.UpdateAllMaturedConsKeyRotatedKeys(ctx, ctx.BlockHeader().Time)
```

Comment:
this gets added a queue, should we just fail and say wait?

Reply:
didn't get the comment, 
this runs in endblocker